vignettes/ggplot2-in-packages.Rmd

```diff
@@ -0,0 +1,247 @@
+
+You will see a lot of other ways to do this in the wild, but the syntax we use here is the only one we can guarantee will work in the future! In particular, don't use `aes_()` or `aes_string()`, as they are deprecated and may be removed in a future version. Finally, don't skip the step of creating a data frame and a mapping to pass in to `ggplot()` or its layers! You will see other ways of doing this in the wild, but these rely on undocumented behaviour and can fail in unexpected ways.
+
+## Best practices for common tasks
+
+### Using ggplot2 to visualize an object
+
+ggplot2 is commonly used in packages to visualize objects (e.g., in a `plot()`-style function). For example, a package might define an S3 object that represents the probability of various discrete values:
+
+```{r}
+mpg_drv_dist <- structure(
+  c(
+    "4" = 103 / 234,
+    "f" = 106 / 234,
+    "r" = 25 / 234
+  ),
+  class = "discrete_distr"
+)
+```
+
+Many S3 objects in R implement a `plot()` method, but it is unrealistic to expect that a single `plot()` method can provide the visualization every one of your users is looking for. It is useful, however, to provide a `plot()` method as a visual summary that users can call to understand the essence of an object. To satisfy all your users, we suggest writing a function that transforms the object into a data frame (or a `list()` of data frames if your object is more complicated). A good example of this approach is [ggdendro](https://cran.r-project.org/package=ggdendro), which creates dendrograms using ggplot2 but also computes the data necessary for users to make their own. For the above example, the function might look like this:
+
+```{r}
+discrete_distr_data <- function(x) {
+  tibble::tibble(
+    value = names(x),
+    probability = as.numeric(x)
+  )
+}
+
+discrete_distr_data(mpg_drv_dist)
+```
+
+In general, users of `plot()` call it for its side-effects: it results in a graphic being displayed. This is different than the behaviour of a `ggplot()`, which is not rendered unless it is explicitly `print()`ed. Because of this, ggplot2 defines its own generic `autoplot()`, a call to which is expected to return a `ggplot()` (with no side effects).
+
+```{r}
+#' @importFrom ggplot2 autoplot
+#' @importFrom rlang .data
+autoplot.discrete_distr <- function(object, ...) {
+  plot_data <- discrete_distr_data(object)
+  ggplot(plot_data, aes(.data$value, .data$probability)) +
+    geom_col() +
+    coord_flip() +
+    labs(x = "Value", y = "Probability")
+}
+```
+
+Once an `autoplot()` method has been defined, a `plot()` method can then consist of `print()`ing the result of `autoplot()`:
+
+```{r}
+#' @importFrom graphics plot
+plot.discrete_distr <- function(x, ...) {
+  print(autoplot(x, ...))
+}
+```
+
+If you don't use ggplot2 for your visualizations but would like to implement `autoplot()` for users that do, it is possible to register your generics only if ggplot2 is installed using `vctrs::s3_register()`.
```

Comment:
This needs to more clearly indicate that it's about adding a method for a ggplot2 generic, if ggplot2 is in suggests (not imports). And generally I'd recommend copying and pasting the source of `vctrs::s3_register` to avoid having to add it as a dependency.